internal/proxyservice/http/aws/aws.go

```diff
@@ -0,0 +1,191 @@
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
+
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector"
+)
+
+// From https://github.com/aws/aws-sdk-go/blob/master/aws/signer/v4/v4.go#L77
+const (
+	timeFormat = "20060102T150405Z"
+)
+
+// newAmzDate parses a date string using the AWS signer time format
+func newAmzDate(amzDateStr string) (time.Time, error) {
+	if amzDateStr == "" {
+		return time.Time{}, fmt.Errorf("missing required header: %s", "X-Amz-Date")
+	}
+
+	return time.Parse(timeFormat, amzDateStr)
+}
+
+// requestMetadataFromAuthz parses an authorization header string and create a
+// requestMetadata instance populated with the associated region and service
+// name
+func requestMetadataFromAuthz(authorizationStr string) (*requestMetadata, error) {
+	// extract credentials section of authorization header
+	credentialsComponent, err := extractCredentialsComponent(authorizationStr)
+	if err != nil {
+		return nil, err
+	}
+	// validate credential component of authorization header, then extract region
+	// and service name
+	matches := headerRegexp.FindStringSubmatch(credentialsComponent)
+	if len(matches) != 3 {
+		return nil, fmt.Errorf("malformed credential component of Authorization header")
+	}
+
+	return &requestMetadata{
+		region:      matches[1],
+		serviceName: matches[2],
+	}, nil
+}
+
+// AWS4-HMAC-SHA256 Credential=AKIAJC5FABNOFVBKRWHA/20171103/us-east-1/ec2/aws4_request
+// See https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+var headerRegexp = regexp.MustCompile(`^Credential=\w+\/\d+\/([\w-_]+)\/(\w+)\/aws4_request$`)
+
+// requestMetadata captures the metadata of a signed AWS request: date, region
+// and service name
+type requestMetadata struct {
+	date        time.Time
+	region      string
+	serviceName string
+}
+
+// extractCredentialsComponent extracts the credentials component from an
+// authorization header string
+func extractCredentialsComponent(authorizationStr string) (string, error) {
+	// Parse the following (line breaks added):
+	// AWS4-HMAC-SHA256
+	// Credential=AKIAJC5FABNOFVBKRWHA/20171103/us-east-1/ec2/aws4_request, \
+	// SignedHeaders=content-type;host;x-amz-date, \
+	// Signature=c4a8ade09a5e0c644cc282311c36aae6c834596076ffde7db7d1195c7b454ed0
+
+	// validate form of entire authorization header
+	tokens := strings.Split(authorizationStr, ", ")
+	if len(tokens) != 3 || tokens[0] == "" || tokens[1] == "" || tokens[2] == "" {
+		return "", fmt.Errorf("malformed Authorization header")
+	}
+
+	// trim prefix and return credential component
+	return strings.TrimPrefix(tokens[0], "AWS4-HMAC-SHA256 "), nil
+}
+
+// newRequestMetadata parses the request headers to extract the metadata
+// necessary to sign the request
+func newRequestMetadata(r *gohttp.Request) (*requestMetadata, error) {
+	authorizationStr := strings.Join(r.Header["Authorization"], "")
+	amzDateStr := strings.Join(r.Header["X-Amz-Date"], "")
```

Comment:
Could you do, eg, `r.Header["Authorization"][0]` instead?  It seems like that would be a more direct way of saying "we expect there to be only one value here" (assuming that is the case...)

Reply:
headers can be an array. this strategy concatenates them as opposed to assuming one value.